File: app/src/main/java/com/mxt/anitrend/base/custom/sheet/BottomSheetList.java
package com.mxt.anitrend.base.custom.sheet;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.Log;
import android.view.View;

import com.mxt.anitrend.R;
import com.mxt.anitrend.base.custom.recycler.RecyclerViewAdapter;
import com.mxt.anitrend.base.custom.recycler.StatefulRecyclerView;
import com.mxt.anitrend.base.custom.view.container.CustomSwipeRefreshLayout;
import com.mxt.anitrend.base.custom.viewmodel.ViewModelBase;
import com.mxt.anitrend.base.interfaces.event.ItemClickListener;
import com.mxt.anitrend.base.interfaces.event.RecyclerLoadListener;
import com.mxt.anitrend.util.CompatUtil;
import com.mxt.anitrend.util.KeyUtil;
import com.nguyenhoanglam.progresslayout.ProgressLayout;

import java.util.List;

import butterknife.BindView;

public abstract class BottomSheetList<T extends Parcelable> extends BottomSheetBase<List<T>> implements ItemClickListener<T>, Observer<List<T>>, RecyclerLoadListener, CustomSwipeRefreshLayout.OnRefreshAndLoadListener {

    protected List<T> model;

    protected @BindView(R.id.stateLayout) ProgressLayout stateLayout;
    protected @BindView(R.id.recyclerView) StatefulRecyclerView recyclerView;

    protected RecyclerViewAdapter<T> mAdapter;
    protected StaggeredGridLayoutManager mLayoutManager;

    protected int mColumnSize;
    protected boolean isPager, isLimit;

    private final View.OnClickListener stateLayoutOnClick = view -> {
        stateLayout.showLoading();
        onRefresh();
    };

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if(getArguments() != null)
            model = getArguments().getParcelableArrayList(KeyUtil.arg_list_model);
        setViewModel(true);
    }

    @Override
    public void onStart() {
        super.onStart();
        stateLayout.showLoading();
        if(model == null)
            onRefresh();
        else
            updateUI();
    }

    protected void addScrollLoadTrigger() {
        if(isPager)
            if (!recyclerView.hasOnScrollListener()) {
                presenter.initListener(mLayoutManager, this);
                recyclerView.addOnScrollListener(presenter);
            }
    }

    protected void removeScrollLoadTrigger() {
        if (isPager)
            recyclerView.clearOnScrollListeners();
    }

    @Override
    public void onPause() {
        super.onPause();
        addScrollLoadTrigger();
    }

    @Override
    public void onResume() {
        removeScrollLoadTrigger();
        super.onResume();
    }

    /**
     * Set your adapter and call this method when you are done to the current
     * parents data, then call this method after
     */
    protected void injectAdapter() {
        mAdapter.setClickListener(this);
        if (recyclerView.getAdapter() == null) {
            recyclerView.setHasFixedSize(true);
            recyclerView.setNestedScrollingEnabled(true);
            recyclerView.setLayoutManager(mLayoutManager);
            recyclerView.setAdapter(mAdapter);
        }
        else
            mAdapter.onItemsInserted(model);
        if (mAdapter.getItemCount() < 1)
            stateLayout.showEmpty(CompatUtil.getDrawable(getContext(),  R.drawable.ic_new_releases_white_24dp, R.color.colorStateBlue), getString(R.string.layout_empty_response));
        else
            stateLayout.showContent();
    }

    /**
     * Is automatically called in the @onStart Method if overridden in list implementation
     */
    protected abstract void updateUI();

    @SuppressWarnings("unchecked")
    protected void setViewModel(boolean stateSupported) {
        if(viewModel == null) {
            viewModel = ViewModelProviders.of(this).get(ViewModelBase.class);
            viewModel.setContext(getContext());
            if(!viewModel.getModel().hasActiveObservers())
                viewModel.getModel().observe(this, this);
            if(stateSupported)
                viewModel.setState(this);
        }
    }

    /**
     * While paginating if our request was a success and
     */
    public void setLimitReached() {
        if(presenter != null && presenter.getCurrentPage() != 0)
            isLimit = true;
    }

    @Override
    public void onRefresh() {
        model = null;
        if(mAdapter != null)
            mAdapter.clearItems();
        if (isPager && presenter != null)
            presenter.onRefreshPage();
        makeRequest();
    }

    @Override
    public void onLoad() {

    }

    @Override
    public void onLoadMore() {
        makeRequest();
    }

    /**
     * All new or updated network requests should be handled in this method
     */
    public abstract void makeRequest();

    /**
     * Called when the data is changed.
     *
     * @param data The new data
     */
    @Override
    public void onChanged(@Nullable List<T> data) {
        Log.d(TAG, "onChanged(@Nullable List<T> data) invoked");
    }

    @Override
    public void showError(String error) {
        super.showError(error);
        stateLayout.showLoading();
        stateLayout.showError(CompatUtil.getDrawable(getContext(), R.drawable.ic_emoji_cry),
                error, getString(R.string.try_again), stateLayoutOnClick);
    }

    @Override
    public void showEmpty(String message) {
        super.showEmpty(message);
        stateLayout.showLoading();
        stateLayout.showError(CompatUtil.getDrawable(getContext(), R.drawable.ic_emoji_sweat),
                message, getString(R.string.try_again) , stateLayoutOnClick);
    }
}

File: app/src/main/java/com/mxt/anitrend/base/custom/sheet/BottomSheetList.kt
package com.mxt.anitrend.base.custom.sheet

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.mxt.anitrend.R
import com.mxt.anitrend.base.custom.recycler.RecyclerViewAdapter
import com.mxt.anitrend.base.custom.recycler.StatefulRecyclerView
import com.mxt.anitrend.base.custom.view.container.CustomSwipeRefreshLayout
import com.mxt.anitrend.base.custom.viewmodel.acquireTypedViewModelBase
import com.mxt.anitrend.base.interfaces.event.ItemClickListener
import com.mxt.anitrend.base.interfaces.event.RecyclerLoadListener
import com.mxt.anitrend.extension.getCompatDrawable
import com.mxt.anitrend.widget.ProgressLayout
import timber.log.Timber

abstract class BottomSheetList<T : android.os.Parcelable> :
    BottomSheetBase<List<T>>(),
    ItemClickListener<T>,
    androidx.lifecycle.Observer<List<T>?>,
    RecyclerLoadListener,
    CustomSwipeRefreshLayout.OnRefreshAndLoadListener {
    protected var stateLayout: ProgressLayout? = null
    protected var recyclerView: StatefulRecyclerView? = null

    protected lateinit var mAdapter: RecyclerViewAdapter<T>
    protected lateinit var mLayoutManager: StaggeredGridLayoutManager

    protected var mColumnSize: Int = 0
    protected var isPager: Boolean = false
    protected var isLimit: Boolean = false

    private val stateLayoutOnClick =
        View.OnClickListener {
            stateLayout?.showLoading()
            onRefresh()
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setViewModel(true)
    }

    override fun onStart() {
        super.onStart()
        stateLayout?.showLoading()
        if (mAdapter.itemCount < 1) {
            onRefresh()
        } else {
            updateUI()
        }
    }

    protected fun addScrollLoadTrigger() {
        if (isPager) {
            val recycler = recyclerView ?: return
            if (!recycler.hasOnScrollListener()) {
                presenter.initListener(mLayoutManager, this)
                recycler.addOnScrollListener(presenter)
            }
        }
    }

    protected fun removeScrollLoadTrigger() {
        if (isPager) {
            recyclerView?.clearOnScrollListeners()
        }
    }

    protected fun bindListViews(rootView: View) {
        stateLayout = rootView.findViewById(R.id.stateLayout)
        recyclerView = rootView.findViewById(R.id.recyclerView)
    }

    override fun onPause() {
        removeScrollLoadTrigger()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        addScrollLoadTrigger()
    }

    protected fun injectAdapter() {
        mAdapter.setClickListener(this)
        val recycler = recyclerView ?: return
        if (recycler.adapter == null) {
            recycler.setHasFixedSize(true)
            recycler.isNestedScrollingEnabled = true
            recycler.layoutManager = mLayoutManager
            recycler.adapter = mAdapter
        }
        if (mAdapter.itemCount < 1) {
            val drawable =
                context?.getCompatDrawable(
                    R.drawable.ic_new_releases_white_24dp,
                    R.color.colorStateBlue,
                ) ?: return
            stateLayout?.showEmpty(drawable, getString(R.string.layout_empty_response))
        } else {
            stateLayout?.showContent()
        }
    }

    protected abstract fun updateUI()

    protected fun setViewModel(stateSupported: Boolean) {
        if (viewModel == null) {
            viewModel = acquireTypedViewModelBase(this, stateSupported, this)
        }
    }

    fun setLimitReached() {
        if (presenter.currentPage != 0) {
            isLimit = true
        }
    }

    override fun onRefresh() {
        if (isPager) {
            presenter.onRefreshPage()
        }
        makeRequest()
    }

    override fun onLoad() = Unit

    override fun onLoadMore() {
        makeRequest()
    }

    abstract fun makeRequest()

    override fun onChanged(data: List<T>?) {
        Timber.tag(TAG ?: javaClass.simpleName).d("onChanged(@Nullable List<T> data) invoked")
    }

    override fun showError(error: String) {
        super.showError(error)
        val drawable = context?.getCompatDrawable(R.drawable.ic_emoji_cry)
        stateLayout?.showError(drawable, error, getString(R.string.try_again), stateLayoutOnClick)
    }

    override fun showEmpty(message: String) {
        super.showEmpty(message)
        val drawable = context?.getCompatDrawable(R.drawable.ic_emoji_sweat)
        stateLayout?.showError(drawable, message, getString(R.string.try_again), stateLayoutOnClick)
    }
}
